hack/cloudfront/readme.md

```diff
@@ -0,0 +1,278 @@
+
+# 1000 Foot View
+CloudFront allows serverless functions to be [executed before processing a request itself](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-cloudfront-trigger-events.html). The serverless function can
+modify the request or even respond to the client directly. This means we can write a request handler capable of detecting
+the AWS region from which the client is making a request and redirecting it directly to S3 if we determine, for example,
+that they are an EC2 instance in us-east-1.
+
+So a client from us-east-1 requesting a blob would be processed as follows:
+1. Make a GET request to registry.ci.openshift.org (`oc whoami -t` as the Authorization token).
+2. The registry pod would verify the token and create a presigned URL for the configured CloudFront distribution and send a redirect for the client to the CloudFront URL.
+3. The client receives the redirect and sends a GET to CloudFront. CloudFront verifies the presigned URL.
+4. Our serverless function is invoked by CloudFront and detects that the client IP is from us-east-1.
+5. Our serverless function creates a presigned URL for the registry's S3 storage bucket. It returns the presigned URL to the client as an HTTP redirect.
+6. The client receives the redirect and sends a GET to the S3 HTTP service.
+7. S3 validates the presigned URL and streams the content back to the client.
+Cost: Free except for the low cost of the CloudFront GET request handling and serverless function execution time.
+
+A client from GCP requesting a blob would be processed as follows:
+1. Make a GET request to registry.ci.openshift.org (`oc whoami -t` as the Authorization token).
+2. The registry pod would verify the token and create a presign URL for the configured CloudFront distribution and send a redirect for the client to the CloudFront URL.
+3. The client receives the redirect and sends a GET to CloudFront. CloudFront verifies the presigned URL.
+4. Our serverless function is invoked by CloudFront and detects that the client IP is NOT from AWS and allows the request to pass through to CloudFront.
+5. CloudFront streams the content back to the client (CloudFront reads transparently from S3 when necessary and this read is free).
+Cost: CloudFront GET request and serverless function execution time + Discounted CloudFront transfer cost.
+
+# 100 Foot View
+We can do even better! What if we had a replica of the app.ci registry in every major region (us-east-1, us-east-2, us-west-1, us-west-2)? 
+If we did, then CloudFront could redirect the client to the "free" app.ci S3 replica bucket within its own region. 
+The problem with this is that when a client pushes content registry.ci.openshift.org, it is only written to one S3 bucket (the one is us-east-1). 
+To replicate the data, S3 is set up with a replication job that will copy content to the other regions within 15 minutes.
```

Comment:
Is this it?
https://aws.amazon.com/getting-started/hands-on/replicate-data-using-amazon-s3-replication/

Could we link it if that is the case?